src/Conversion/ONNXToTOSA/Tensor/Resize.cpp

```diff
@@ -0,0 +1,213 @@
+//
+// This file lowers the ONNX Resize operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include <cstdint>
+#include <numeric>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXResizeOpLoweringToTOSA : public ConversionPattern {
+public:
+  ONNXResizeOpLoweringToTOSA(MLIRContext *ctx)
+      : ConversionPattern(ONNXResizeOp::getOperationName(), 1, ctx) {}
+  using OpAdaptor = typename ONNXResizeOp::Adaptor;
+
+  /// ## coordinateTransformationMode ##
+  /// TOSA uses the formula ix = (ox * scale_x_d + offset_x) / scale_x_n
+  /// to find the input coordinates. In order to lower ONNX one needs to
+  /// express the modes in this context. Border is always
+  /// border = d * (output - 1) - n * (input - 1) + offset;
+  ///
+  /// ### half_pixel ###
+  /// ONNX formula: ix = (ox + 0.5) / scale - 0.5
+  /// gcd = greatest common divisor
+  /// To meet TOSA requirements:
+  /// - scale_x_d = input_size * 2 / gcd
+  /// - scale_x_n = output_size * 2 / gcd
+  /// - offset_x =  (scale_x_d - scale_x_n) / 2
+  ///
+  /// ### pytorch_half_pixel ###
+  /// Same as half_pixel, but if output == 1:
+  /// - scale_x_d = scale_x_n = 1
+  /// - offset = -1
+  ///
+  /// ### align_corners ###
+  /// - scale_x_d = (input_size - 1)
+  /// - scale_x_n = (output_size - 1)
+  /// - offset = 0
+  ///
+  /// ### asymmetric ###
+  /// - scale_x_d = input_size
+  /// - scale_x_n = output_size
+  /// - offset = 0
+  ///
+  /// ## nearest_mode ##
+  /// If mode == nearest, then ONNX can set the nearest_mode attribute.
+  /// The standard case for TOSA of round_half_up. Support for floor can
+  /// be achieved with:
+  /// offset_x -= n/2
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    auto resizeOp = llvm::cast<ONNXResizeOp>(op);
+    Location loc = op->getLoc();
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+
+    Value input = adaptor.X();
+    auto inputType = input.getType().dyn_cast<RankedTensorType>();
+
+    auto resultType =
+        resizeOp.getResult().getType().dyn_cast<RankedTensorType>();
+
+    StringRef mode = adaptor.mode();
+    StringRef nearestMode = adaptor.nearest_mode();
+    StringRef coordinateTransformationMode =
+        adaptor.coordinate_transformation_mode();
+
+    if (inputType.getRank() != 4) {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA only support 4D tensors as input of resize.");
+    }
+
+    if (inputType.isDynamicDim(2) || inputType.isDynamicDim(3)) {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "Only static sized tensors are supported.");
+    }
```

Comment:
Add a comment afterwards to say that `roi`, `scales` and `sizes` are therefore not present/relevant.